synapseutils/migrate_functions.py

```diff
@@ -0,0 +1,736 @@
+
+
+def _record_file_migration(cursor, entity_id, mapping):
+    # record all the file handle migrations associated with a FileEntity migration
+    # (can be multiple of multiple versions where migrated)
+    insert_values = []
+    for version, file_handle_ids in mapping.items():
+        from_file_handle_id, to_file_handle_id = file_handle_ids
+        insert_values.append((entity_id, version, from_file_handle_id, to_file_handle_id))
+
+    cursor.executemany(
+        """
+            insert into file_entity_versions (id, version, from_file_handle_id, to_file_handle_id)
+            values (?, ?, ?, ?)
+        """,
+        insert_values
+    )
+
+
+def _record_table_migration(cursor, entity_id, mapping):
+    # record all the file handle migrations associated with a TableEntity migration
+    # (can be multiple if there were multiple rows/columns of file handles)
+    insert_values = []
+    for row_id, col_mapping in mapping.items():
+        for col_name, file_handle_ids in col_mapping.items():
+            from_file_handle_id, to_file_handle_id = file_handle_ids
+            insert_values.append((entity_id, row_id, col_name, from_file_handle_id, to_file_handle_id))
+
+    cursor.executemany(
+        """
+            insert into table_entity_files (id, row, column, from_file_handle_id, to_file_handle_id)
+            values (?, ?, ?, ?, ?)
+        """,
+        insert_values
+    )
+
+
+def _entity_type_of(entity):
+    concrete_type = utils.concrete_type_of(entity)
+
+    # e.g. 'org.sagebionetworks.repo.model.FileEntity' -> 'file'
+    concrete_type_short = concrete_type.rsplit('.', 1)[-1].lower().replace('entity', '')
+    return concrete_type_short
+
+
+def _check_indexed(cursor, entity):
+    # check if we have indexed the given entity in the sqlite db yet.
+    # if so it can skip reindexing it. supports resumption.
+    entity_id = utils.id_of(entity)
+    indexed_row = cursor.execute(
+        "select 1 from entities where id = ? and status >= ?",
+        (entity_id, _MigrationStatus.INDEXED.value)
+    ).fetchone()
+
+    if indexed_row:
+        logging.info('%s already indexed, skipping', entity_id)
+        return True
+
+    logging.debug('%s not yet indexed, indexing now', entity_id)
+    return False
+
+
+def _index_entity(conn, cursor, syn, entity, parent_id):
+    # recursive function to index a given entity into the sqlite db.
+
+    entity_id = utils.id_of(entity)
+
+    if not _check_indexed(cursor, entity_id):
+        entity_type = _entity_type_of(entity)
+
+        if entity_type == 'file' or entity_type == 'table':
+            cursor.execute(
+                'insert into entities (id, parent_id, type, status) values (?, ?, ?, ?)',
+                (entity_id, parent_id, entity_type, _MigrationStatus.INDEXED.value)
+            )
+
+        elif entity_type == 'folder' or entity_type == 'project':
+            children = syn.getChildren(entity_id, includeTypes=['file', 'table', 'folder'])
```

Comment:
Don't walk tables for now... just FileEntities

Reply:
By default we only walk files now, but tables can be included with the optional include_table_files param.